### PR TITLE
Add some improvements to the Gatsby Open Preview button

### PR DESCRIPTION
**Pull request by abhiaiyer91**
We can fetch the exact path e.g. `blog/post` from the preview instance, so attempt to fetch or fallback to instance parameters.

This requires gatsby 2.4.7 users using prior versions will get a CORS error and the behavior will fallback to the original instance parameter method

### CODE REVIEW

File: samples/gatsby-preview/src/index.js [hunk 1]
```diff
@@ -68,8 +55,45 @@ class App extends React.Component {
...
+          variables: {
+            slugExpr: `/${contentSlug ? contentSlug.getValue() : contentTypeSlug}\/?$/`
+          }
+        })
+      })
+        .then(res => res.json())
+        .then(json => (json && json.data ? json.data.sitePage.path : ``));
+
+      const normalize = part => part.replace(/\/$/, '');
+      window.open(`${normalize(previewUrl)}/${slug}`);
```
Comment (DSchau) [inline samples/gatsby-preview/src/index.js]: Note: future optimization here--fail or prompt if slug not found?

**Comment by DSchau** [comment]
Hello!

This is a fairly substantial improvement for some of our Gatsby Preview customers (using Contentful!), so we'd love a review on this so we can get it in their hands.

Looks like @floelhoeffel has been requested--let either @abhiaiyer91 or myself know if we can help with anything!

**Comment by floelhoeffel** [comment]
Sorry for the delay guys. I get an engineer to review before end of week.

File: samples/gatsby-preview/src/index.js [hunk 2]
```diff
@@ -68,8 +55,45 @@ class App extends React.Component {
...
   };
 
-  openPreviewTab = () => {
-    window.open(this.state.previewUrl);
+  openPreviewTab = async () => {
+    const {
+      parameters: { installation, instance },
+      entry
+    } = this.props.sdk;
+    const { previewUrl } = installation;
```
Comment (jelz) [inline samples/gatsby-preview/src/index.js]: I'd normalize this URL right away - you're using normalized version in L85 but not normalized in L68.

File: samples/gatsby-preview/src/index.js [hunk 3]
```diff
@@ -68,8 +55,45 @@ class App extends React.Component {
...
+        },
+        body: JSON.stringify({
+          query:
+            'query getQualifiedSlug($slugExpr:String) { sitePage( path:{ regex:$slugExpr } ) { path } }',
+          variables: {
+            slugExpr: `/${contentSlug ? contentSlug.getValue() : contentTypeSlug}\/?$/`
+          }
+        })
+      })
+        .then(res => res.json())
```
Comment (jelz) [inline samples/gatsby-preview/src/index.js]: Don't mix `await` and `then`, I'd expect to see:

```js
const res = await fetch(/* ... */);
const { data } = await res.json();
const slug = data ? data.sitePage.path : '';
```

File: samples/gatsby-preview/src/index.js [hunk 4]
```diff
@@ -68,8 +55,45 @@ class App extends React.Component {
...
+        .then(json => (json && json.data ? json.data.sitePage.path : ``));
+
+      const normalize = part => part.replace(/\/$/, '');
+      window.open(`${normalize(previewUrl)}/${slug}`);
+    } catch (e) {
+      console.error(e);
+
+      let slug = contentTypeSlug ? contentTypeSlug : '';
```
Comment (jelz) [inline samples/gatsby-preview/src/index.js]: It's equivalent to `let slug = contentTypeSlug || '';`

File: samples/gatsby-preview/src/index.js [hunk 5]
```diff
@@ -68,8 +55,45 @@ class App extends React.Component {
...
+
+      if (this.props.sdk.entry.fields.slug) {
```
Comment (jelz) [inline samples/gatsby-preview/src/index.js]: you could use `contentSlug` instead of `this.props.sdk.entry.fields.slug`

**Comment by DSchau** [comment]
@jelz addressed your feedback; thanks!

**Comment by DSchau** [comment]
@jelz ✅ to your feedback; thanks!

File: samples/gatsby-preview/src/index.js [hunk 6]
```diff
@@ -68,8 +57,46 @@ class App extends React.Component {
...
+      entry
+    } = this.props.sdk;
+    const { previewUrl } = installation;
+    const { contentTypeSlug } = instance;
+    const { slug: contentSlug } = entry.fields;
+
+    const normalizedPreviewUrl = normalize(previewUrl)
+
+    try {
+      const res = await fetch(`${previewUrl}/___graphql`, {
```
Comment (jelz) [inline samples/gatsby-preview/src/index.js]: how about using `normalizedPreviewUrl` here too?

Comment (DSchau) [reply samples/gatsby-preview/src/index.js]:
```suggestion
      const res = await fetch(`${normalizedPreviewUrl}/___graphql`, {
```

**Comment by jelz** [comment]
@DSchau one last small comment.

To answer your question from the commit message - yes, you still need to commit the built bundle. We're thinking about automation here.

**Comment by DSchau** [comment]
@jelz @abhiaiyer91 will test soon when he gets a chance--but this seems good to go from my end. Thanks!

**Comment by jelz** [comment]
ok, just ping me when done testing

**Comment by abhiaiyer91** [comment]
@jelz everything is running great! We're ready! Thanks everyone!

**Pull request merged**